README
Copyright (C) Manuel A. Morales

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

A copy of the GNU General Public License should be available along
with the R documentation. If not, write to the Free Software Foundation,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA, or download it
from the internet.

              http://www.gnu.org/copyleft/gpl.html

                     -------------------

This package contains the sciplot library for R.

Authors: Manuel Morales <mmorales@williams.edu> with code from
the R Core Team.


CHANGELOG
1.2-0 Fix for for default axes labels. Remove option trace.label from
      lineplot.CI. TODO: Subset fix in 1.1-1 does not work using with()
      unless $ is used for the subset variable.
1.1-1 Fix to honor subset argument when data option is not specified and
      added details to NAMESPACE file
1.1-0 Fix to remove unused levels from group factors. Fix to change
      lwd in lineplot.CI when no group factors are specified.
1.0-9 Fix to remove unused levels from factors. Change default
      ordering of factor levels in legend when more than one group factor.
1.0-8 Fix to honor factor levels with more than one group factor
1.0-7 Fix to default ylim when all negative values in bargraph.CI
1.0-6 Fix for legend code with fixed=TRUE in lineplot.CI
1.0-5 Fix for incomplete factorial designs in bargraph.CI
1.0-4 Remove period from GPL license
1.0-3 Fix for labeling legend with "fixed=TRUE" option
1.0-2 Fix to enable "las" parameter in lineplot.CI
1.0-1 Fix for unbalanced designs.
1.0-0 Add ability to plot data with non-overlapping x-axis factors in
      lineplot.CI.
0.9-0 Remove functions (xppauto.plot, dev2ps) not associated with factorial
      graphs to package separately.
0.8-2 Add options for specifiying width, color and linetype of error bars.
0.8-1 Add example data file for xppauto.plot(). Upload package to CRAN.
0.8-0 Add xppauto.plot for plotting bifurcation diagrams from XPPAUT and
      utility function dev2ps.
0.7-0 Change trace.factor to group. Add option to create a "split plot" for the
      bargraph function similar to the appearance of a back-to-back histogram.
0.6-1 Fix bug with CI's in lineplot.CI().
0.6-0 Modified CI to allow asymmetric functions. Changed handling of data
      environment.
0.5-0 First release.


R/bargraphCI.R
bargraph.CI <-
    function(x.factor, response, group = NULL, split = FALSE,
             xlab=NULL, ylab=NULL, col = NULL, angle = NULL, density = NULL,
             lc = TRUE, uc = TRUE, legend = FALSE, ncol = 1,
             leg.lab = NULL, x.leg = NULL, y.leg = NULL, cex.leg = 1,
             bty = "n", bg = "white", space=if(split) c(-1,1),
             err.width=if(length(levels(as.factor(x.factor)))>10) 0 else .1,
             err.col = "black", err.lty = 1,
             fun = function(x) mean(x, na.rm = TRUE),
             ci.fun = function(x) c(fun(x)-se(x), fun(x)+se(x)),
             ylim = NULL, xpd = FALSE, data = NULL, subset = NULL, ...) {
  
        ## Set up environment
        if(is.null(xlab)) xlab <- deparse(substitute(x.factor))
        if(is.null(ylab)) ylab <-deparse(substitute(response))
        
        subset <- eval(substitute(subset), envir=data)
        
        if(!is.null(data)) {
            if(!is.null(subset)) data <- subset(data,subset)
            x.factor <- eval(substitute(x.factor), envir=data)
            response <- eval(substitute(response), envir=data)
            group <- eval(substitute(group), envir=data)
        }
        
        if(is.null(data)) {
            if(!is.null(subset)) {
                x.factor <- x.factor[subset]
                response <- response[subset]
                group <- group[subset]
            }
        }
        
        subset = NULL

        if(split) {
            ## If more than 1 "y-factor", Return Error Message
            if(length(group[[1]]) > 1) {
                print("Error: Can't split for > 2 group levels")
                stop()
            }

            split.fn <- function(x, y) {
                val <- levels(group)[[1]]
                ifelse(x==val, y*-1, y)
            }

            response <- split.fn(group, response)
        }

        ## Figure out if we're dealing with 1 or more total groups
        ## (i.e., we always have 1 "x.factor" and we may have 1 or more "group".
        if(is.null(group)) groups = factor(x.factor) else {
            ## If more than 1 "y-factor", combine for plotting
            if(length(group[[1]]) > 1) {
                group <- factor(interaction(group, lex.order=TRUE))
            }
            ## "groups" defines the combination of "x.factor" and "group"
            group <- factor(group) # Do this to drop unused levels
            groups <- list(group,x.factor)
        }
        
        ## Calculate mean and SE's
        mn.data <- tapply(response, groups, fun)
        CI.dat <- tapply(response, groups, ci.fun)

        ## replace NULL with NaN
        null.fn <- function(x) if(is.null(x[[1]])) rep(NaN,2) else x
        if(!is.null(group)) CI.dat <- apply(CI.dat, c(1,2), null.fn)
        CI.data <-
            array(unlist(CI.dat),
                  c(2,
                    if(is.null(group)) 1 else nrow(mn.data),
                    length(levels(as.factor(x.factor)))))
        CI.L <- CI.data[1,,]
        CI.H <- CI.data[2,,]
        ## Replace undefined SE with zero.
        ##Note that this will return the warning message:
        ## "zero-length arrow is of indeterminate angle and so skipped"
        replace.NA <- function(x) if(is.na(x)) 0 else x
        
        if(!is.null(group)) {
            CI.L <- apply(CI.L, c(1,2), replace.NA)
            CI.H <- apply(CI.H, c(1,2), replace.NA)
        }
        else {
            CI.L <- as.vector(unlist(lapply(CI.L, replace.NA)))
            CI.H <- as.vector(unlist(lapply(CI.H, replace.NA)))
        }

        ## Determine y-axis plot region
        if(is.null(ylim)) ylim <- c(min(0, CI.L), max(0, CI.H))
        
        ## Plot
        xvals <- barplot(mn.data, ylim=ylim, beside = TRUE, col = col,
                         density=density, angle = angle, space=space,
                         xlab=xlab, ylab=ylab, xpd=xpd, ...)
        ## Draw CI's
        nlevels.x <- dim(mn.data)[1]
        if(is.null(group))
            arrows(xvals, if(lc) CI.L else mn.data, xvals,
                   if(uc) CI.H else mn.data, angle=90, length=err.width, code=3)
        else {
            nlevels.y <- dim(mn.data)[2]
            for(i in 1:nlevels.y)
                arrows(xvals[,i], if(lc) CI.L[,i] else mn.data[,i],
                       xvals[,i], if(uc) CI.H[,i] else mn.data[,i],
                       angle=90, code = 3, col=err.col, lty=err.lty,
                       length=err.width)
        }

        ## Draw legend
        if(!is.null(group)&legend) {
            legend(
                x = if(is.null(x.leg)) 0.8*max(xvals) else x.leg,
                y = if(is.null(y.leg)) max(ylim) else y.leg,
                legend = if(is.null(leg.lab))
                             levels(as.factor(group)) else leg.lab,
                bty = bty, bg = bg, ncol = ncol,
                fill = if(is.null(col)) gray.colors(nlevels.x) else col,
                density = density, angle = angle, cex=cex.leg)
        }
        
        invisible(list(xvals=xvals, vals=mn.data, CI=CI.data))
    }


R/bargraphCI2.R


bargraph.CI <-
  function(x.factor, response, group = NULL, split = FALSE,
           col = NULL, angle = NULL, density = NULL,
           lc = TRUE, uc = TRUE, legend = FALSE, ncol = 1,
           leg.lab = NULL, x.leg = NULL, y.leg = NULL, cex.leg = 1,
           bty = "n", bg = "white", space=if(split) c(-1,1),
           err.width=if(length(levels(as.factor(x.factor)))>10) 0 else .1,
           err.col = "black", err.lty = 1,
           fun = function(x) mean(x, na.rm = TRUE),
           ci.fun = function(x) c(fun(x)-se(x), fun(x)+se(x)),
           ylim = NULL, xpd = FALSE, data = NULL, subset = NULL, ...) {
  
    # Set up environment
    subset <- eval(substitute(subset), envir=data)
    
    if(!is.null(data)) {
      if(!is.null(subset)) data <- subset(data,subset)
      x.factor <- eval(substitute(x.factor), envir=data)
      response <- eval(substitute(response), envir=data)
      group <- eval(substitute(group), envir=data)
    }

    subset = NULL

    if(split) {
      # If more than 1 "y-factor", Return Error Message
      if(length(group[[1]]) > 1) {
        print("Error: Can't split for > 2 group levels")
        stop()
      }

      split.fn <- function(x, y) {
        val <- levels(group)[[1]]
        ifelse(x==val, y*-1, y)
      }

      response <- split.fn(group, response)
    }

    # Figure out if we're dealing with 1 or more total groups (i.e., we always
    # have 1 "x.factor" and we may have 1 or more "group".
    if(is.null(group)) groups = x.factor else {
      # If more than 1 "y-factor", combine for plotting
      if(length(group[[1]]) > 1) {
        group <- do.call("paste",c(group,sep="."))
      }
      # "groups" defines the combination of "x.factor" and "group"
      groups <- list(group,x.factor)
    }

    # Calculate mean and SE's
    mn.data <- tapply(response, groups, fun)
    CI.dat <- tapply(response, groups, ci.fun)

    ## replace NULL with NaN
    null.fn <- function(x) if(is.null(x[[1]])) rep(NaN,2) else x
    if(!is.null(group)) CI.dat <- apply(CI.dat, c(1,2), null.fn)
    CI.data <-
      array(unlist(CI.dat),
            c(2,
              if(is.null(group)) 1 else nrow(mn.data),
              length(levels(as.factor(x.factor)))))
    CI.L <- CI.data[1,,]
    CI.H <- CI.data[2,,]
    # Replace undefined SE with zero. Note that this will return the warning
    # message: "zero-length arrow is of indeterminate angle and so skipped"
    replace.NA <- function(x) if(is.na(x)) 0 else x

    if(!is.null(group)) {
      CI.L <- apply(CI.L, c(1,2), replace.NA)
      CI.H <- apply(CI.H, c(1,2), replace.NA)
    }
    else {
      CI.L <- as.vector(unlist(lapply(CI.L, replace.NA)))
      CI.H <- as.vector(unlist(lapply(CI.H, replace.NA)))
    }

    # Determine y-axis plot region
    if(is.null(ylim)) ylim <- c(min(0, CI.L), max(CI.H))

    # Plot
    xvals <- barplot(mn.data, ylim=ylim, beside = TRUE, col = col,
                     density=density, angle = angle, space=space,
                     xpd=xpd, ...)
    # Draw CI's
    nlevels.x <- dim(mn.data)[1]
    if(is.null(group))
      arrows(xvals, if(lc) CI.L else mn.data, xvals,
             if(uc) CI.H else mn.data, angle=90, length=err.width, code=3)
    else {
      nlevels.y <- dim(mn.data)[2]
      for(i in 1:nlevels.y)
        arrows(xvals[,i], if(lc) CI.L[,i] else mn.data[,i],
               xvals[,i], if(uc) CI.H[,i] else mn.data[,i],
               angle=90, code = 3, col=err.col, lty=err.lty,
               length=err.width)
    }

    # Draw legend
    if(!is.null(group)&legend) {
      legend(
             x = if(is.null(x.leg)) 0.8*max(xvals) else x.leg,
             y = if(is.null(y.leg)) max(ylim) else y.leg,
             legend = if(is.null(leg.lab))
             levels(as.factor(group)) else leg.lab,
             bty = bty, bg = bg, ncol = ncol,
             fill = if(is.null(col)) gray.colors(nlevels.x) else col,
             density = density, angle = angle, cex=cex.leg)
    }

    invisible(list(xvals=xvals, vals=mn.data, CI=CI.data))
  }


R/lineplot.CI.R
lineplot.CI <-
  function(x.factor, response, group = NULL, type = "b", x.cont=FALSE,
           legend = TRUE, trace.label = NULL, leg.lab = NULL,
           fixed = FALSE, x.leg = NULL, y.leg = NULL, cex.leg = 1, ncol = 1,
           pch = c(16,21,15,22,17,24, c(3:14)),
           fun = function(x) mean(x, na.rm=TRUE),
           ci.fun = function(x) c(fun(x)-se(x), fun(x)+se(x)),
           err.width = if(length(levels(as.factor(x.factor)))>10) 0 else .1,
           err.col = col, err.lty = 1, xlim = NULL, ylim = NULL,
           cex = NULL, lwd = NULL, col = "black", cex.axis = 1,
           xaxt = "s", data = NULL, subset = NULL, ...) {
  # Set up environment
  subset <- eval(substitute(subset), envir=data)
  fun <- eval(substitute(fun), envir=data)
  if(!is.null(data)) {
    if(!is.null(subset)) data <- subset(data,subset)
    x.factor <- eval(substitute(x.factor), envir=data)
    response <- eval(substitute(response), envir=data)
    group <- eval(substitute(group), envir=data)
  }

  subset = NULL
  
  #////////////////////////////////////////////////////////////////////#
  # Below modified from the interaction.plot function in stats package #
  # Take the legend section out to allow greater flexibility           #
  #////////////////////////////////////////////////////////////////////#

  int.plot <- function (x.factor = x.factor, group = group,
                        response = response, type = c("l","p", "b"),
                        legend = legend,
                        trace.label = deparse(substitute(group)),
                        fixed = FALSE, xlab = deparse(substitute(x.factor)),
                        ylab = ylabel, lty = nc:1, pch = NA, xpd = NULL,
                        leg.bg = par("bg"), leg.bty = "n", xtick = FALSE,
                        xlim=xlim, ylim=ylim, axes = TRUE, ...) {
  
    ylabel <- paste(deparse(substitute(fun)), "of ", 
                    deparse(substitute(response)))
    type <- match.arg(type)
    cells <- tapply(response, list(x.factor, group), fun)
    nr <- nrow(cells)
    nc <- ncol(cells)
    xvals <- if(x.cont) as.numeric(levels(as.factor(x.factor))) else 1:nr
    if (is.ordered(x.factor)) {
      wn <- getOption("warn")
      options(warn = -1)
      xnm <- as.numeric(levels(x.factor))
      options(warn = wn)
      if (!any(is.na(xnm)))
        xvals <- xnm
    }
    xlabs <- rownames(cells)
    ylabs <- colnames(cells)
    nch <- max(sapply(ylabs, nchar, type = "width"))
    if (is.null(xlabs))
      xlabs <- as.character(xvals)
    if (is.null(ylabs))
      ylabs <- as.character(1:nc)
    if(is.null(xlim)) { 
      xlim <- range(xvals)
      xleg <- xlim[2] + 0.05 * diff(xlim)
      xlim <- xlim + c(-0.2/nr, if (legend&is.null(x.leg)&is.null(y.leg))
                       0.2 + 0.02 * nch else 0.2/nr) * diff(xlim)}
    else {
      xlim
      xleg <- xlim[2] - 0.25 * diff(xlim)}
    matplot(xvals, cells, ..., type = type, xlim = xlim, ylim = ylim,
            xlab = xlab, ylab = ylab, axes = axes, xaxt = "n", col = col,
            lty = lty, lwd=lwd, pch = pch)
    if (axes && xaxt != "n") {
      axisInt <- function(x, main, sub, lwd, bg, log, asp,
                          ...) axis(1, x, ...)
      mgp. <- par("mgp")
      if (!xtick)
        mgp.[2] <- 0
      axisInt(1, at = xvals, labels = xlabs, tick = xtick,
              mgp = mgp., xaxt = xaxt, ...)
    }
    ord <- sort.list(cells[nr, ], decreasing = TRUE)
    if (legend) {
      yrng <- diff(ylim)
      yleg <- ylim[2] - 0.1 * yrng
      if (!is.null(xpd) || {
        xpd. <- par("xpd")
        !is.na(xpd.) && !xpd. && (xpd <- TRUE)
      }) {
        op <- par(xpd = xpd)
        on.exit(par(op))
      }
      text(xleg, ylim[2] - 0.05 * yrng, paste("  ", trace.label),
           adj = 0)
      if (!fixed) {
        ylabs <- ylabs[ord]
        lty <- lty[1 + (ord - 1)%%length(lty)]
        col <- col[1 + (ord - 1)%%length(col)]
        pch <- pch[ord]
      }
    }
    invisible()

    return.data<-if(legend)
      list(pch=pch,ord=ord,xleg=xleg,yleg=yleg,ylabs=ylabs,lty=lty,
           leg.bty=leg.bty,leg.bg=leg.bg,ord=ord,xvals=xvals,cells=cells)
    else list(pch=pch,ord=ord,xvals=xvals)
       return(return.data)
  }
  #////////////////////////////////////////////////////////////////////#
  #                         End Section                                #
  #////////////////////////////////////////////////////////////////////#

  # Figure out if we're dealing with 1 or more total groups (i.e., we always
  # have 1 "x.factor" and we may have 1 or more "group".
  if(is.null(group)) groups = factor(x.factor) else {   
    # If more than 1 "y-factor", combine for plotting
    if(length(group[[1]]) > 1)
      group <- factor(interaction(group, lex.order=TRUE))
    # "groups" defines the combination of "x.factor" and "group"
    groups <- list(x.factor, group)
  }

  # Calculate mean and SE's
  mn.data <- tapply(response, groups, fun)
  CI.data <- tapply(response, groups, ci.fun)
  # Determine y-axis plot region
  plot.limits = c(
    min(c(unlist(mn.data), unlist(CI.data)), na.rm=TRUE),
    max(c(unlist(mn.data), unlist(CI.data)), na.rm=TRUE))

  # Draw lines
  if(is.null(group)) {
    nlevels.x <-
      if(x.cont) as.numeric(levels(as.factor(x.factor)))
      else 1:nrow(mn.data)
    plot(nlevels.x, mn.data, xaxt="n", type=type, col=col, pch=NA, cex=cex,
         cex.axis=cex.axis,
         xlim=if(is.null(xlim)) {
           c(min(nlevels.x)-.2,max(nlevels.x)+.2)
         }
         else xlim,
         ylim=if(is.null(ylim)) plot.limits else ylim, ...)
    if(xaxt!="n") axis(1,label=names(mn.data),
         at=nlevels.x, cex.axis=cex.axis, ...)
  }
  else leg.vals <-
    int.plot(x.factor, group, response, type = type,
             xlim = xlim, ylim = if(is.null(ylim)) plot.limits else ylim,
             cex.axis = cex.axis, trace.label = trace.label, pch = NA,
             legend = legend, ...)
  # Draw CI's
  if(is.null(group)) {
    nlevels.x <-
      if(x.cont) as.numeric(levels(as.factor(x.factor)))
      else 1:nrow(mn.data)
    CI.seln <- !is.na(mn.data)
    CI.plot <- matrix(unlist(CI.data[CI.seln]),
                      nrow=sum(CI.seln), byrow=TRUE)
    arrows(nlevels.x[CI.seln],
           CI.plot[,1],
           nlevels.x[CI.seln],
           CI.plot[,2],
           angle = 90, col = err.col, length = err.width, code = 3,
           lwd = lwd, lty=err.lty)
  }
  else {
    nlevels.y <- ncol(mn.data)
    for(i in 1:nlevels.y) {
      CI.seln <- !is.na(mn.data)[,i]
      CI.plot <- matrix(unlist(CI.data[CI.seln,i]),
                        nrow=sum(CI.seln), byrow=TRUE)
      arrows(leg.vals$xvals[CI.seln],
             CI.plot[,1],
             leg.vals$xvals[CI.seln],
             CI.plot[,2],
             angle=90, length=err.width,
             col=if(length(err.col)>1) err.col[i] else err.col,
             lty=if(length(err.lty)>1) err.lty[i] else err.lty,
             code=3,lwd=lwd)
    }
  }

  # Draw points (Note: adding points at this point allows points to be in
  # the foreground)
  if(type %in% c("p", "b")) {
    if(is.null(group)) {
      nlevels.x <-
        if(x.cont) as.numeric(levels(as.factor(x.factor)))
        else 1:nrow(mn.data)
      points(nlevels.x, mn.data,pch=pch[1],bg="white",cex=cex,col=col)
    }
    else {
      nlevels.y<-dim(mn.data)[2]
      for(i in 1:nlevels.y)
        points(leg.vals$xvals, mn.data[,i],pch=pch[i],bg="white",
               col=if(length(col)>1) col[i] else col,cex=cex)
    }
  }

  #////////////////////////////////////////////////////////////////////#
  # Below modified from the interaction.plot function in stats package #
  #////////////////////////////////////////////////////////////////////#

    if(legend & !is.null(group)) {
    legend(x=if(is.null(x.leg)) leg.vals$xleg else x.leg,
           y=if(is.null(y.leg)) leg.vals$yleg else y.leg,
           legend = if(!is.null(leg.lab)) leg.lab else {
             if(fixed) levels(as.factor(unlist(group))) else leg.vals$ylabs
           },
           pch = if(type %in% c("p", "b")) {
             if(!fixed) pch[leg.vals$ord]
             else pch
             },
           col = if(type %in% c("p", "b")) {
             if(!fixed  & length(col)>1) col[leg.vals$ord]
             else col
             },
           lty = if (type %in% c("l", "b")) {
             if(fixed) leg.vals$lty[order(leg.vals$ord)]
             else leg.vals$lty
             },
           ncol=ncol, bty = leg.vals$leg.bty, bg = leg.vals$leg.bg,
           cex=cex.leg)
  }

  #////////////////////////////////////////////////////////////////////#
  #                      End Section                                   #
  #////////////////////////////////////////////////////////////////////#

  invisible(list(vals=mn.data, CI=CI.data))
}


R/xppauto.plot.R
xppauto.plot <- function(data, xlim=NULL, ylim=NULL,
                         max.step=NULL, min.step=NULL, ...) {
  
  if(is.null(xlim))
    xlim <- c(min(data$V1), max(data$V1))
  if(is.null(ylim))
    ylim <- c(min(data$V2,data$V3), max(c(data$V2,data$V3)))
  if(is.null(max.step))
    max.step <- (xlim[2]-xlim[1])*.075
  if(is.null(min.step))
    min.step <- (xlim[2]-xlim[1])*.0075

  data.ymin <- subset(data, V2>ylim[1]&V3>ylim[1])
  cats <- levels(as.factor(data.ymin$V4))
  
  thin.fn <- function(x) {
      index <- vector()
      index[1] <- 1
      if(length(x$V1)>1) {
        for(i in 2:length(x$V1)) {
          if(abs(x$V1[i]-x$V1[max(index)])>=min.step)
            index[(length(index)+1)] <- i
        }
      }
      x[index,]
    }
  
  breaks.fn <- function(x) {  # Split up "unconnected" vectors
    breaks <- vector()
    breaks[1] <- 1
    for(i in 1:(length(x)-1)) {
      breaks[i+1] <- ifelse(abs(x[i+1]-x[i])>max.step,breaks[i]+1, breaks[i])
      }
    return(breaks)
  }

  plot(xlim, ylim, type="n", ...)

  for(i in 1:length(cats)) {
    data.cat <- subset(data.ymin, V4==cats[[i]])
    breaks <- breaks.fn(data.cat$V1)
    for(j in 1:max(breaks)) {
      data.break <- subset(data.cat, breaks==j)
      data.plot <- if(i<3) data.break else thin.fn(data.break)
      resp1 <- data.plot$V2
      resp2 <- data.plot$V3
      
      points(resp1~V1, data=data.plot, type=ifelse(i<3, "l", "p"),
             lty=if(i==2) 2, pch=if(i==3) 16)
      points(resp2~V1, data=data.plot, type=ifelse(i<3, "l", "p"),
             lty=if(i==2) 2, pch=if(i==3) 16)
    }
  }
  invisible(ylim=ylim)
}


R/lineplotCI.R
lineplot.CI <-
    function(x.factor, response, group = NULL, type = "b", xlab=NULL,
             ylab=NULL, x.cont=FALSE, legend = TRUE, leg.lab = NULL,
             fixed = FALSE, x.leg = NULL, y.leg = NULL, cex.leg = 1,
             ncol = 1, pch = c(16,21,15,22,17,24, c(3:14)),
             fun = function(x) mean(x, na.rm=TRUE),
             ci.fun = function(x) c(fun(x)-se(x), fun(x)+se(x)),
             err.width = if(length(levels(as.factor(x.factor)))>10) 0 else .1,
             err.col = col, err.lty = 1, xlim = NULL, ylim = NULL,
             cex = NULL, lwd = NULL, col = "black", cex.axis = 1,
             xaxt = "s", data = NULL, subset = NULL, ...) {
      
        ## Set up environment
        if(is.null(xlab)) xlab <- deparse(substitute(x.factor))
        if(is.null(ylab)) ylab <-deparse(substitute(response))

        subset <- eval(substitute(subset), envir=data)
        fun <- eval(substitute(fun), envir=data)
        if(!is.null(data)) {
            if(!is.null(subset)) data <- subset(data,subset)
            x.factor <- eval(substitute(x.factor), envir=data)
            response <- eval(substitute(response), envir=data)
            group <- eval(substitute(group), envir=data)
        }

        if(is.null(data)) {
            if(!is.null(subset)) {
                x.factor <- x.factor[subset]
                response <- response[subset]
                group <- group[subset]
            }
        }

        subset = NULL
        
        ##////////////////////////////////////////////////////////////////////#
        ## Below modified from the interaction.plot function in stats package #
        ## Take the legend section out to allow greater flexibility           #
        ##////////////////////////////////////////////////////////////////////#
        
        int.plot <- function (x.factor = x.factor, group = group,
                              response = response, type = c("l","p", "b"),
                              legend = legend,
                              fixed = FALSE, lty = nc:1, pch = NA, xpd = NULL,
                              leg.bg = par("bg"), leg.bty = "n", xtick = FALSE,
                              xlim=xlim, ylim=ylim, axes = TRUE, ...) {
            
            type <- match.arg(type)
            cells <- tapply(response, list(x.factor, group), fun)
            nr <- nrow(cells)
            nc <- ncol(cells)
            xvals <- if(x.cont) as.numeric(levels(as.factor(x.factor)))
                     else 1:nr
            if (is.ordered(x.factor)) {
                wn <- getOption("warn")
                options(warn = -1)
                xnm <- as.numeric(levels(x.factor))
                options(warn = wn)
                if (!any(is.na(xnm)))
                    xvals <- xnm
            }
            xlabs <- rownames(cells)
            ylabs <- colnames(cells)
            nch <- max(sapply(ylabs, nchar, type = "width"))
            if (is.null(xlabs))
                xlabs <- as.character(xvals)
            if (is.null(ylabs))
                ylabs <- as.character(1:nc)
            if(is.null(xlim)) { 
                xlim <- range(xvals)
                xleg <- xlim[2] + 0.05 * diff(xlim)
                xlim <-
                    xlim + c(-0.2/nr,
                             if (legend&is.null(x.leg)&is.null(y.leg))
                                 0.2 + 0.02 * nch else 0.2/nr) * diff(xlim)}
            else {
                xlim
                xleg <- xlim[2] - 0.25 * diff(xlim)}
            matplot(xvals, cells, ..., type = type, xlim = xlim, ylim = ylim,
                    xlab = xlab, ylab = ylab, axes = axes, xaxt = "n",
                    col = col, lty = lty, lwd=lwd, pch = pch)
            if (axes && xaxt != "n") {
                axisInt <- function(x, main, sub, lwd, bg, log, asp, ...)
                    axis(1, x, ...)
                mgp. <- par("mgp")
                if (!xtick)
                    mgp.[2] <- 0
                axisInt(1, at = xvals, labels = xlabs, tick = xtick,
                        mgp = mgp., xaxt = xaxt, ...)
            }
            ord <- sort.list(cells[nr, ], decreasing = TRUE)
            if (legend) {
                yrng <- diff(ylim)
                yleg <- ylim[2] - 0.1 * yrng
                if (!is.null(xpd) || {
                    xpd. <- par("xpd")
                    !is.na(xpd.) && !xpd. && (xpd <- TRUE)
                }) {
                    op <- par(xpd = xpd)
                    on.exit(par(op))
                }
                if (!fixed) {
                    ylabs <- ylabs[ord]
                    lty <- lty[1 + (ord - 1)%%length(lty)]
                    col <- col[1 + (ord - 1)%%length(col)]
                    pch <- pch[ord]
                }
            }
            invisible()

            return.data <-
                if(legend)
                    list(pch=pch,ord=ord,xleg=xleg,yleg=yleg,ylabs=ylabs,
                         lty=lty,leg.bty=leg.bty,leg.bg=leg.bg,ord=ord,
                         xvals=xvals,cells=cells)
                else list(pch=pch,ord=ord,xvals=xvals)
            return(return.data)
        }
        ##////////////////////////////////////////////////////////////////////#
        ##                         End Section                                #
        ##////////////////////////////////////////////////////////////////////#

        ## Figure out if we're dealing with 1 or more total groups
        ## (i.e., we always have 1 "x.factor" and we may have 1 or more "group".
        if(is.null(group)) groups = factor(x.factor) else {   
            ## If more than 1 "y-factor", combine for plotting
            if(length(group[[1]]) > 1)
                group <- factor(interaction(group, lex.order=TRUE))
            ## "groups" defines the combination of "x.factor" and "group"
            group <- factor(group) # Do this to drop unused levels
            groups <- list(x.factor, group)
        }

        ## Calculate mean and SE's
        mn.data <- tapply(response, groups, fun)
        CI.data <- tapply(response, groups, ci.fun)
        ## Determine y-axis plot region
        plot.limits = c(
            min(c(unlist(mn.data), unlist(CI.data)), na.rm=TRUE),
            max(c(unlist(mn.data), unlist(CI.data)), na.rm=TRUE))

        ## Draw lines
        if(is.null(group)) {
            nlevels.x <-
                if(x.cont) as.numeric(levels(as.factor(x.factor)))
                else 1:nrow(mn.data)
            plot(nlevels.x, mn.data, xaxt="n", type=type, col=col, pch=NA,
                 cex=cex, cex.axis=cex.axis, lwd=lwd, xlab=xlab, ylab=ylab,
                 xlim=if(is.null(xlim)) {
                     c(min(nlevels.x)-.2,max(nlevels.x)+.2)
                 }
                      else xlim,
                 ylim=if(is.null(ylim)) plot.limits else ylim, ...)
            if(xaxt!="n") axis(1,labels=names(mn.data),
                   at=nlevels.x, cex.axis=cex.axis, ...)
        }
        else leg.vals <-
            int.plot(x.factor, group, response, type = type,
                     xlim = xlim, ylim = if(is.null(ylim)) plot.limits
                                         else ylim,
                     cex.axis = cex.axis, pch = NA, legend = legend, ...)
        ## Draw CI's
        if(is.null(group)) {
            nlevels.x <-
                if(x.cont) as.numeric(levels(as.factor(x.factor)))
                else 1:nrow(mn.data)
            CI.seln <- !is.na(mn.data)
            CI.plot <- matrix(unlist(CI.data[CI.seln]),
                              nrow=sum(CI.seln), byrow=TRUE)
            arrows(nlevels.x[CI.seln],
                   CI.plot[,1],
                   nlevels.x[CI.seln],
                   CI.plot[,2],
                   angle = 90, col = err.col, length = err.width, code = 3,
                   lwd = lwd, lty=err.lty)
        }
        else {
            nlevels.y <- ncol(mn.data)
            for(i in 1:nlevels.y) {
                CI.seln <- !is.na(mn.data)[,i]
                CI.plot <- matrix(unlist(CI.data[CI.seln,i]),
                                  nrow=sum(CI.seln), byrow=TRUE)
                arrows(leg.vals$xvals[CI.seln],
                       CI.plot[,1],
                       leg.vals$xvals[CI.seln],
                       CI.plot[,2],
                       angle=90, length=err.width,
                       col=if(length(err.col)>1) err.col[i] else err.col,
                       lty=if(length(err.lty)>1) err.lty[i] else err.lty,
                       code=3,lwd=lwd)
            }
        }

        ## Draw points (Note: adding points at this point allows points to be in
        ## the foreground)
        if(type %in% c("p", "b")) {
            if(is.null(group)) {
                nlevels.x <-
                    if(x.cont) as.numeric(levels(as.factor(x.factor)))
                    else 1:nrow(mn.data)
                points(nlevels.x, mn.data,pch=pch[1],bg="white",cex=cex,col=col)
            }
            else {
                nlevels.y<-dim(mn.data)[2]
                for(i in 1:nlevels.y)
                    points(leg.vals$xvals, mn.data[,i],pch=pch[i],bg="white",
                           col=if(length(col)>1) col[i] else col,cex=cex)
            }
        }
        
        ##////////////////////////////////////////////////////////////////////#
        ## Below modified from the interaction.plot function in stats package #
        ##////////////////////////////////////////////////////////////////////#

        if(legend & !is.null(group)) {
            legend(x=if(is.null(x.leg)) leg.vals$xleg else x.leg,
                   y=if(is.null(y.leg)) leg.vals$yleg else y.leg,
                   legend = if(!is.null(leg.lab)) leg.lab else {
                       if(fixed) levels(as.factor(unlist(group)))
                       else leg.vals$ylabs
                   },
                   pch = if(type %in% c("p", "b")) {
                       if(!fixed) pch[leg.vals$ord]
                       else pch
                   },
                   col = if(type %in% c("p", "b")) {
                       if(!fixed  & length(col)>1) col[leg.vals$ord]
                       else col
                   },
                   lty = if (type %in% c("l", "b")) {
                       if(fixed) leg.vals$lty[order(leg.vals$ord)]
                       else leg.vals$lty
                   },
                   ncol=ncol, bty = leg.vals$leg.bty, bg = leg.vals$leg.bg,
                   cex=cex.leg)
        }

        ##////////////////////////////////////////////////////////////////////#
        ##                      End Section                                   #
        ##////////////////////////////////////////////////////////////////////#

        invisible(list(vals=mn.data, CI=CI.data))
    }
